### DISCUSSION TITLE
SignalR hub method, handling "incorrect parameters" exceptions

### DISCUSSION THREAD

**Discussion by TimeWanderer** [dotnet/aspnetcore#29171] [Q&A]
Is it possible to catch a hub message parse error and take custom action? From what I gather, these exceptions do not reach any hub filters either.
The client gets an error message "Error: Failed to invoke 'hubMethod' due to an error on the server." when there is a parse error on MessagePack or JSON. There are many cases we want to validate the messages and return more meaningful error messages to the client.

**Comment by BrennanConroy** [comment]
> Is it possible to catch a hub message parse error and take custom action?

Nope, these are most often development time errors and are likely uncommon in production.

> There are many cases we want to validate the messages and return more meaningful error messages to the client.

If there is a parsing error then there wont be anything for you to validate, just a `byte[]`. Can you describe your scenario(s) and how you would want it to behave?

**Comment by TimeWanderer** [comment]
"Validate" might be the wrong terminology. I am sorry for that. A simple case for me goes like this:

I have a hub method that is used by both our own applications and customer's applications. When the method parameter is invalid, the client gets a generic message.

This is not a development time exception for us, as we have a dynamic query model as a parameter to this hub method. I would like to return a more informative error message, for example, which part of the parameter object is erroneous. 

When using JsonSerializer, and parsing the parameter fails, server-side logs an error with debug level:

_Parameters to hub method 'methodName' are incorrect.
System.IO.InvalidDataException: Error binding arguments. Make sure that the types of the provided values match the types of the hub method being invoked.
       ---> System.Text.Json.JsonException: The JSON value could not be converted to System.Int32. Path: $.view.limit | LineNumber: 0 | BytePositionInLine: 1186.
       ---> System.InvalidOperationException: Cannot get the value of a token type 'Null' as a number._

I have two problems here:
- Log level is debug, so I have no way of monitoring these errors on production apart from setting some very specific config. This point might be invalid, because of the logging guidelines of asp.net core in general, of which I am not aware. Probably setting Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher's log level to Debug from appsettings.json would solve this issue for me, albeit with some extra logs.

- I have no way of intercepting parse exception. This error occurs before hub filters. I could not figure out a solution to this problem.

I think the ability to override ErrorMessageHelper behaviour to send clients customized error messages would be great for my current case and many other cases. Current behaviour is set by "includeExceptionDetails" hub configuration. 

P.S. I am aware this might be a very specific issue, I am open to any suggestions/alternatives.
